src/Components/Components/src/Routing/Router.cs

```diff
@@ -155,12 +181,55 @@ private void Refresh(bool isNavigationIntercepted)
             }
         }
 
+        private async Task RunOnNavigateAsync(string path, bool isNavigationIntercepted)
+        {
+            // If this router instance does not provide an OnNavigateAsync parameter
+            // then we render the component associated with the route as per usual.
+            if (OnNavigateAsync == null)
+            {
+                Refresh(isNavigationIntercepted);
+                return;
+            }
+
+            // Create a new CTS for this invocation of the task and
+            // extract the reference to the current one.
+            var pendingOnNavigateCts = _onNavigateCts;
+            var onNavigateCts = new CancellationTokenSource();
+            _onNavigateCts = onNavigateCts;
+
+            var navigateContext = new OnNavigateArgs(path, _onNavigateCts);
+            var task = OnNavigateAsync(navigateContext);
+
+            // If we've already invoked a task and stored its CTS, then
+            // cancel the existing task.
+            if (pendingOnNavigateCts != null)
+            {
+                pendingOnNavigateCts.Cancel();
```

Comment:
Was there a reason to defer cancelling the previous CT until after calling `OnNavigate`? It might be surprising to developers that when `OnNavigate` is called, the previous CT hasn't yet been cancelled.

Also the implementation would be simpler if cancelling the previous one was move up to around line 196 (no need for a `pendingOnNavigateCts` variable).

Reply:
Yep -- good call. I think this particular ordering is a result of me mucking around with different approaches.